src/validation.cpp

```diff
@@ -4872,3 +4906,329 @@ class CMainCleanup
+    m_snapshot_metadata = MakeUnique<SnapshotMetadata>(metadata);
+    return true;
+}
+
+std::vector<CChainState*> ChainstateManager::GetAll()
+{
+    std::vector<CChainState*> out;
+
+    if (!IsSnapshotValidated() && m_ibd_chainstate) {
+        out.push_back(m_ibd_chainstate.get());
+    }
+
+    if (m_snapshot_chainstate) {
+        out.push_back(m_snapshot_chainstate.get());
+    }
+
+    return out;
+}
+
+void ChainstateManager::RunOnAll(
+    const std::function<void(CChainState*)> fn)
+{
+    for (CChainState* chainstate : GetAll()) {
+        fn(chainstate);
+    }
+}
+
+std::vector<CChainState*> ChainstateManager::GetAllForBlockDownload()
+{
+    std::vector<CChainState*> out;
+
+    if (m_snapshot_chainstate) out.push_back(m_snapshot_chainstate.get());
+    if (!IsSnapshotValidated() && m_ibd_chainstate) out.push_back(m_ibd_chainstate.get());
+
+    return out;
+}
+
+CChainState* ChainstateManager::InitializeChainstate(
+    size_t cache_size_bytes,
+    bool in_memory,
+    bool should_wipe,
+    bool activate,
+    const uint256& snapshot_blockhash,
+    bool snapshot_needs_initialization)
+{
+    // If we don't yet have metadata for the snapshot, we need to do some extra processing
+    // before it's ready for use, so stash it in `m_unready_snapshot_chainstate` to be
+    // processed in ActivateSnapshot().
+    std::unique_ptr<CChainState>& to_modify = (
+        snapshot_blockhash.IsNull() ? m_ibd_chainstate :
+            snapshot_needs_initialization ? m_unready_snapshot_chainstate : m_snapshot_chainstate);
+
+    to_modify.reset(new CChainState(
+        cache_size_bytes, in_memory, should_wipe, "chainstate", snapshot_blockhash));
+
+    if (activate) {
+        LogPrintf("Switching active chainstate to %s\n", snapshot_blockhash.ToString());
+        m_active_chainstate = to_modify.get();
+    }
+
+    return to_modify.get();
+}
+
+bool ChainstateManager::ActivateSnapshot(
+        CAutoFile* coins_file,
+        SnapshotMetadata metadata,
+        size_t cache_size_bytes,
+        bool in_memory)
+{
+    uint256 base_blockhash = metadata.m_base_blockheader.GetBlockHash();
+
+    // Important that we haven't set `m_snapshot_metadata` yet before this point.
+    CChainState* snapshot_chainstate = InitializeChainstate(
+        cache_size_bytes, in_memory,
+        /* should_wipe */ false, /* activate */ false,
+        /* snapshot_blockhash */ base_blockhash, /* snapshot_needs_initialization */ true);
+
+    CCoinsViewCache* coins_cache = snapshot_chainstate->CoinsCache();
+
+    CCoinsMap coins_map;
+    COutPoint outpoint;
+    Coin coin;
+    uint64_t coins_count = metadata.m_coins_count;
+    uint64_t coins_left = metadata.m_coins_count;
+
+    LogPrintf("[snapshot] loading coins from snapshot %s\n", base_blockhash.ToString());
+
+    while (coins_left > 0) {
+        *coins_file >> outpoint;
+        *coins_file >> coin;
+        coins_map[outpoint] = CCoinsCacheEntry(std::move(coin), CCoinsCacheEntry::DIRTY);
```

Comment:
Note: this needs to be rewritten to avoid assuming that the user has enough free memory to fit the entire snapshot. Right now it'll OOM on machines with less than 3.2GB mem available.